src/java/org/apache/cassandra/index/internal/CassandraIndexSearcher.java

```diff
@@ -90,17 +106,32 @@ private ClusteringIndexFilter makeIndexFilter(ReadCommand command)
     {
         if (command instanceof SinglePartitionReadCommand)
         {
-            // Note: as yet there's no route to get here - a 2i query *always* uses a
-            // PartitionRangeReadCommand. This is here in preparation for coming changes
-            // in SelectStatement.
             SinglePartitionReadCommand sprc = (SinglePartitionReadCommand)command;
             ByteBuffer pk = sprc.partitionKey().getKey();
             ClusteringIndexFilter filter = sprc.clusteringIndexFilter();
 
             if (filter instanceof ClusteringIndexNamesFilter)
             {
-                NavigableSet<Clustering<?>> requested = ((ClusteringIndexNamesFilter)filter).requestedRows();
-                BTreeSet<Clustering<?>> clusterings = BTreeSet.copy(requested, index.getIndexComparator());
+                if (index instanceof CollectionValueIndex)
+                {
+                    // Collection value indexes have an extra clustering key for the path, but we cannot construct an
+                    // index names filter from the filter on the backing table, because it has no path information.
+                    // Instead, we construct a slice from the clustering keys that are provided...
+                    Slices slices = filter.getSlices(index.baseCfs.metadata());
+                    ClusteringBound<?> start = BufferClusteringBound.BOTTOM;
+                    ClusteringBound<?> end = BufferClusteringBound.TOP;
+
+                    if (slices.size() > 0)
+                        start = slices.get(0).start();
+                    if (slices.size() > 0)
+                        end = slices.get(slices.size() - 1).end();
+
+                    Slice slice = Slice.make(makeIndexBound(pk, start), makeIndexBound(pk, end));
+                    return new ClusteringIndexSliceFilter(Slices.with(index.getIndexComparator(), slice), filter.isReversed());
+                }
```

Comment:
Hopefully the comment explains what's going on here. Is there a more compact way to handle `ClusteringIndexNamesFilter`?